### PR TITLE
[Scheduler] Make sure handlers have synced before scheduling

### LINKED ISSUE #116717: Scheduler: Make sure handlers have synced before scheduling
### What would you like to be added?

Make sure handlers have finished syncing before the scheduling cycles start.

Ref: https://github.com/kubernetes/kubernetes/issues/113763#issuecomment-1472858978

/sig scheduling
/good-first-issue

### Why is this needed?

In a highly used cluster, we don't want to start scheduling pods before all the pods have been loaded into the scheduler cache, or we could end up in a bad state.

### PR DESCRIPTION
**Pull request by AxeZhan**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Make sure handlers have finished syncing before the scheduling cycles start.
#### Which issue(s) this PR fixes:

Fixes #116717

#### Special notes for your reviewer:
This may not be a clean approach because it changes the return parameters for `scheduler.New`, and I am looking forward to finding some more clean solutions.
#### Does this PR introduce a user-facing change?

```release-note
 schedular now waits for handlers to finish syncing before the scheduling cycles start.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by kidddddddddddddddddddddd** [comment]
Current implementation returns a `registeredHandlers []cache.ResourceEventHandlerRegistration` for `schedular.New`, and pass this slice all the way to `app.Run`.
I think another option is to add a field `registeredHandlers []cache.ResourceEventHandlerRegistration` for schedular, and code changes will be cleaner, but will cost some more memories.
/cc @alculquicondor for some advice.

**Comment by kidddddddddddddddddddddd** [comment]
> Current implementation returns a `registeredHandlers []cache.ResourceEventHandlerRegistration` for `schedular.New`, and pass this slice all the way to `app.Run`. I think another option is to add a field `registeredHandlers []cache.ResourceEventHandlerRegistration` for schedular, and code changes will be cleaner, but will cost some more memories. /cc @alculquicondor for some advice.

/cc @alculquicondor

**Comment by kidddddddddddddddddddddd** [comment]
> Current implementation returns a `registeredHandlers []cache.ResourceEventHandlerRegistration` for `schedular.New`, and pass this slice all the way to `app.Run`. I think another option is to add a field `registeredHandlers []cache.ResourceEventHandlerRegistration` for schedular, and code changes will be cleaner, but will cost some more memories. /cc @alculquicondor for some advice.

On a second thought, since we never break app.Run(). These handlers will never be garbage collected, so I've decided to use the later option.

File: pkg/scheduler/eventhandlers.go [hunk 1]
```diff
@@ -246,6 +249,34 @@ func responsibleForPod(pod *v1.Pod, profiles profile.Map) bool {
 	return profiles.HandlesSchedulerName(pod.Spec.SchedulerName)
 }
 
+const (
+	// syncedPollPeriod controls how often you look at the status of your sync funcs
+	syncedPollPeriod = 100 * time.Millisecond
+)
+
+// WaitForHandlersSync waits for EventHandlers to sync.
+// It returns true if it was successful, false  if the controller should shut down
```
Comment (denkensk) [inline pkg/scheduler/eventhandlers.go:258]:
```suggestion
// It returns true if it was successful, false if the controller should shut down
```

File: pkg/scheduler/eventhandlers.go [hunk 2]
```diff
@@ -246,6 +249,34 @@ func responsibleForPod(pod *v1.Pod, profiles profile.Map) bool {
...
+func (sched *Scheduler) WaitForHandlersSync(ctx context.Context) bool {
+	err := wait.PollImmediateUntil(syncedPollPeriod,
+		func() (bool, error) {
+			for _, registration := range sched.RegisteredHandlers {
+				if !registration.HasSynced() {
+					return false, nil
+				}
```
Comment (denkensk) [inline pkg/scheduler/eventhandlers.go:265]:
```suggestion
			for _, handler := range sched.RegisteredHandlers {
				if !handler.HasSynced() {
					return false, nil
				}
```

File: pkg/scheduler/eventhandlers.go [hunk 3]
```diff
@@ -255,7 +286,7 @@ func addAllEventHandlers(
 	gvkMap map[framework.GVK]framework.ActionType,
 ) {
 	// scheduled pod cache
-	informerFactory.Core().V1().Pods().Informer().AddEventHandler(
+	resourceEventHandlerRegistration, _ := informerFactory.Core().V1().Pods().Informer().AddEventHandler(
```
Comment (denkensk) [inline pkg/scheduler/eventhandlers.go:289]: To make it clean?
```suggestion
	reg, _ := informerFactory.Core().V1().Pods().Informer().AddEventHandler(
```

File: pkg/scheduler/eventhandlers.go [hunk 4]
```diff
@@ -281,8 +312,10 @@ func addAllEventHandlers(
 			},
 		},
 	)
+	sched.RegisteredHandlers = append(sched.RegisteredHandlers, resourceEventHandlerRegistration)
```
Comment (denkensk) [inline pkg/scheduler/eventhandlers.go:315]: Append  `sched.RegisteredHandlers` again and again makes the code a bit redundant.  😓

Or we can change like  this? WDYT? 
```go
handles := []ResourceEventHandlerRegistration{}
.....
handlers = append(handles, reg)
......
sched.RegisteredHandlers = handlers
```

File: pkg/scheduler/scheduler.go [hunk 5]
```diff
@@ -96,6 +96,9 @@ type Scheduler struct {
 	percentageOfNodesToScore int32
 
 	nextStartNodeIndex int
+
+	// RegisteredHandlers saves registration handle for the handlers to tell if the handler is synced
```
Comment (denkensk) [inline pkg/scheduler/scheduler.go:100]:
```suggestion
	// RegisteredHandlers contains the registrations of all handlers. It's used to check if all handlers have finished syncing before the scheduling cycles start.
```

**Review by denkensk — COMMENTED** [review]
Thanks, @kidddddddddddddddddddddd 
Some comments.

File: pkg/scheduler/eventhandlers.go (hunk 4 above)
Comment (kidddddddddddddddddddddd) [reply pkg/scheduler/eventhandlers.go:315]: Yea, it's a bit redundant, but I didn't find a clean way😢.  We can try your option first.

File: pkg/scheduler/eventhandlers.go [hunk 6]
```diff
@@ -246,6 +249,34 @@ func responsibleForPod(pod *v1.Pod, profiles profile.Map) bool {
...
+				}
+			}
+			return true, nil
+		},
+		ctx.Done())
+
+	if err != nil {
+		klog.FromContext(ctx).V(2).Error(err, "error occurred during WaitForHandlersSync")
+		return false
+	}
```
Comment (denkensk) [inline pkg/scheduler/eventhandlers.go:274]:
```suggestion
	if err := wait.PollImmediateUntil(syncedPollPeriod,
		func() (bool, error) {
			for _, handler := range sched.RegisteredHandlers {
				if !handler.HasSynced() {
					return false, nil
				}
			}
			return true, nil
		},
		ctx.Done()); err != nil {
		klog.FromContext(ctx).V(2).Error(err, "error occurred during WaitForHandlersSync")
		return false
	}
```

**Comment by denkensk** [comment]
lgtm overall.  @alculquicondor  Plz takes a look again.

File: pkg/scheduler/eventhandlers.go [hunk 7]
```diff
@@ -255,7 +284,7 @@ func addAllEventHandlers(
 	gvkMap map[framework.GVK]framework.ActionType,
 ) {
 	// scheduled pod cache
-	informerFactory.Core().V1().Pods().Informer().AddEventHandler(
+	resourceEventHandlerRegistration, _ := informerFactory.Core().V1().Pods().Informer().AddEventHandler(
```
Comment (alculquicondor) [inline pkg/scheduler/eventhandlers.go:287]: Too long. `handlerRegistration` should be enough.

Comment (alculquicondor) [reply pkg/scheduler/eventhandlers.go:287]: Also, do not silence the error, return it.

File: pkg/scheduler/scheduler.go (hunk 5 above)
Comment (alculquicondor) [reply pkg/scheduler/scheduler.go:100]: This is not resolved

File: pkg/scheduler/scheduler.go [hunk 8]
```diff
@@ -96,6 +96,9 @@ type Scheduler struct {
...
+	RegisteredHandlers []cache.ResourceEventHandlerRegistration
```
Comment (alculquicondor) [inline pkg/scheduler/scheduler.go:101]: it doesn't look like this needs to be exported. Rename to `registeredHandlers`.

**Comment by AxeZhan** [comment]
Sry, something went wrong and I didn't fix the review last time(maybe because I forgot to push to origin...)
Updated in new commit.

File: pkg/scheduler/eventhandlers.go [hunk 9]
```diff
@@ -255,11 +255,11 @@ const (
 )
 
 // WaitForHandlersSync waits for EventHandlers to sync.
-// It returns true if it was successful, false  if the controller should shut down
+// It returns true if it was successful, false if the controller should shut down
 func (sched *Scheduler) WaitForHandlersSync(ctx context.Context) bool {
```
Comment (alculquicondor) [inline pkg/scheduler/eventhandlers.go:259]: obtain the log from the context here. We shouldn't be querying the context in each iteration.

File: pkg/scheduler/eventhandlers_test.go [hunk 10]
```diff
@@ -443,7 +443,9 @@ func TestAddAllEventHandlers(t *testing.T) {
 			dynclient := dyfake.NewSimpleDynamicClient(scheme)
 			dynInformerFactory := dynamicinformer.NewDynamicSharedInformerFactory(dynclient, 0)
 
-			addAllEventHandlers(&testSched, informerFactory, dynInformerFactory, tt.gvkMap)
+			if err := addAllEventHandlers(&testSched, informerFactory, dynInformerFactory, tt.gvkMap); err != nil {
+				t.Errorf("Add event handlers failed, error = %v", err)
```
Comment (alculquicondor) [inline pkg/scheduler/eventhandlers_test.go:447]: Fatalf

File: pkg/scheduler/scheduler.go [hunk 11]
```diff
@@ -349,7 +349,9 @@ func New(client clientset.Interface,
 	}
 	sched.applyDefaultHandlers()
 
-	addAllEventHandlers(sched, informerFactory, dynInformerFactory, unionedGVKs(clusterEventMap))
+	if err = addAllEventHandlers(sched, informerFactory, dynInformerFactory, unionedGVKs(clusterEventMap)); err != nil {
+		return nil, fmt.Errorf("add eventhandlers: %v", err)
```
Comment (alculquicondor) [inline pkg/scheduler/scheduler.go:353]:
```suggestion
		return nil, fmt.Errorf("adding event handlers: %w", err)
```

File: pkg/scheduler/eventhandlers.go (hunk 9 above)
Comment (AxeZhan) [reply pkg/scheduler/eventhandlers.go:259]: Like this?
```
func (sched *Scheduler) WaitForHandlersSync(ctx context.Context) bool {
	logger := klog.FromContext(ctx)
```

**Comment by alculquicondor** [comment]
Flaky test #117013

File: pkg/scheduler/eventhandlers.go (hunk 9 above)
Comment (AxeZhan) [reply pkg/scheduler/eventhandlers.go:259]: It's already updated

File: pkg/scheduler/eventhandlers.go [hunk 12]
```diff
@@ -246,16 +249,47 @@ func responsibleForPod(pod *v1.Pod, profiles profile.Map) bool {
...
+		func() (bool, error) {
+			for _, handler := range sched.registeredHandlers {
+				if !handler.HasSynced() {
+					return false, nil
+				}
+			}
+			return true, nil
+		},
+		ctx.Done()); err != nil {
+		logger.V(2).Error(err, "error occurred during WaitForHandlersSync")
```
Comment (alculquicondor) [inline pkg/scheduler/eventhandlers.go:271]: on a second thought: it makes more sense to return the error and let the caller log

And you can make the caller also log `Handlers synced`

Comment (AxeZhan) [reply pkg/scheduler/eventhandlers.go]: Updated. Also, in our use case, I think this error can only be a timeout error. Should we only log the message without `err`?
```
logger.V(2).Error(nil, "timeout during WaitForHandlersSync")
```

File: cmd/kube-scheduler/app/server.go [hunk 13]
```diff
@@ -207,7 +207,11 @@ func Run(ctx context.Context, cc *schedulerserverconfig.CompletedConfig, sched *
 	if cc.DynInformerFactory != nil {
 		cc.DynInformerFactory.WaitForCacheSync(ctx.Done())
 	}
-
+	// Wait for all handlers to sync (all items in the initial list delivered) before scheduling.
+	if err := sched.WaitForHandlersSync(ctx); err != nil {
+		logger.V(2).Error(err, "error occurred during WaitForHandlersSync")
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/server.go:212]: In general, errors should not have a verbosity level.

```suggestion
		logger.Error(err, "Failed waiting for handlers to sync")
```

Comment (alculquicondor) [reply cmd/kube-scheduler/app/server.go]: You missed the correction in the message. Messages in log lines (unlike errors) start with capital https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#remove-string-formatting-from-log-message:~:text=Start%20from%20a%20capital%20letter.

File: test/integration/util/util.go [hunk 14]
```diff
@@ -96,8 +96,13 @@ func StartScheduler(ctx context.Context, clientSet clientset.Interface, kubeConf
 		klog.Fatalf("Error creating scheduler: %v", err)
 	}
 
+	logger := klog.FromContext(ctx)
 	informerFactory.Start(ctx.Done())
 	informerFactory.WaitForCacheSync(ctx.Done())
+	if err = sched.WaitForHandlersSync(ctx); err != nil {
+		logger.Error(err, "error occurred during WaitForHandlersSync")
```
Comment (alculquicondor) [inline test/integration/util/util.go:103]: we shouldn't continue with the test in this case https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#replacing-fatal-calls

Comment (AxeZhan) [reply test/integration/util/util.go]: Ah, I see. Didn't know this before, updated.

Comment (AxeZhan) [reply test/integration/util/util.go]: Should we continue in `app/server.go` when failed waiting for handlers to sync?

File: cmd/kube-scheduler/app/server.go [hunk 15]
```diff
@@ -207,7 +207,11 @@ func Run(ctx context.Context, cc *schedulerserverconfig.CompletedConfig, sched *
 	if cc.DynInformerFactory != nil {
 		cc.DynInformerFactory.WaitForCacheSync(ctx.Done())
 	}
-
+	// Wait for all handlers to sync (all items in the initial list delivered) before scheduling.
+	if err := sched.WaitForHandlersSync(ctx); err != nil {
+		logger.Error(err, "Failed waiting for handlers to sync")
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/server.go:212]: right, we should return the error here:
```suggestion
		return fmt.Errorf("waiting for handlers to sync: %w", err)
```

File: pkg/scheduler/schedule_one_test.go [hunk 16]
```diff
@@ -428,6 +428,9 @@ func TestSchedulerMultipleProfilesScheduling(t *testing.T) {
 	// Run scheduler.
 	informerFactory.Start(ctx.Done())
 	informerFactory.WaitForCacheSync(ctx.Done())
+	if err = sched.WaitForHandlersSync(ctx); err != nil {
+		t.Fatal(err)
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one_test.go:432]:
```suggestion
		t.Fatalf("Handlers failed to sync: %v: ", err)
```

File: test/integration/util/util.go [hunk 17]
```diff
@@ -98,6 +98,11 @@ func StartScheduler(ctx context.Context, clientSet clientset.Interface, kubeConf
 
 	informerFactory.Start(ctx.Done())
 	informerFactory.WaitForCacheSync(ctx.Done())
```
Comment (alculquicondor) [inline test/integration/util/util.go:100]: get a logger from context here

Comment (AxeZhan) [reply test/integration/util/util.go:103]: I added a logger above. But it seems we can't use `ErrorS` with logger

Comment (alculquicondor) [reply test/integration/util/util.go:103]: yeah, just Error

**Review by alculquicondor — COMMENTED** [review]
/approve
/lgtm

File: cmd/kube-scheduler/app/server.go [hunk 18]
```diff
@@ -207,7 +207,11 @@ func Run(ctx context.Context, cc *schedulerserverconfig.CompletedConfig, sched *
 	if cc.DynInformerFactory != nil {
 		cc.DynInformerFactory.WaitForCacheSync(ctx.Done())
 	}
-
+	// Wait for all handlers to sync (all items in the initial list delivered) before scheduling.
+	if err := sched.WaitForHandlersSync(ctx); err != nil {
+		logger.Error(err, "error occurred during WaitForHandlersSync")
```
Comment (aojea) [inline cmd/kube-scheduler/app/server.go]: do we keep going despite it fails?

File: test/integration/util/util.go (hunk 14 above)
Comment (aojea) [reply test/integration/util/util.go]: it seems this has been resolved but I can't see how this stops , logger.Error AFAIK doesn't exit

File: cmd/kube-scheduler/app/server.go (hunk 18 above)
Comment (AxeZhan) [reply cmd/kube-scheduler/app/server.go]: > do we keep going despite it fails?

Nope, we return the error if it fails here. Plz see the new commit.

File: pkg/scheduler/eventhandlers.go [hunk 19]
```diff
@@ -246,16 +249,41 @@ func responsibleForPod(pod *v1.Pod, profiles profile.Map) bool {
...
 
+const (
+	// syncedPollPeriod controls how often you look at the status of your sync funcs
+	syncedPollPeriod = 100 * time.Millisecond
+)
+
+// WaitForHandlersSync waits for EventHandlers to sync.
+// It returns true if it was successful, false if the controller should shut down
+func (sched *Scheduler) WaitForHandlersSync(ctx context.Context) error {
+	return wait.PollImmediateUntil(syncedPollPeriod,
```
Comment (aojea) [inline pkg/scheduler/eventhandlers.go:260]: should this be `wait.PollImmediateUntilWithContext`
```suggestion
	return wait.PollImmediateUntilWithContext(ctx, syncedPollPeriod,
```

Comment (AxeZhan) [reply pkg/scheduler/eventhandlers.go:260]: Both function seem to be deprecated.
```
// Deprecated: This method does not return errors from context, use PollWithContextCancel.
// Note that the new method will no longer return ErrWaitTimeout and instead return errors
// defined by the context package. Will be removed in a future release.
func PollImmediateUntilWithContext(ctx context.Context, interval time.Duration, condition ConditionWithContextFunc) error {
	return poll(ctx, true, poller(interval, 0), condition)
}

```
I found #116712 during writing this pr. Can you confirm `PollImmediateUntilWithContext` is the right function?

Comment (alculquicondor) [reply pkg/scheduler/eventhandlers.go:260]: use PollWithContextCancel

Comment (AxeZhan) [reply pkg/scheduler/eventhandlers.go:260]: > use PollWithContextCancel

Yeah, the thing is there is no function named `PollWithContextCancel` now. #116712

Comment (aojea) [reply pkg/scheduler/eventhandlers.go:260]: yeah, good catch

Comment (AxeZhan) [reply pkg/scheduler/eventhandlers.go:260]: So, should I use `PollImmediateUntilWithContext` in this pr or merge this first?
I think scheduler is using `wait.Poll` in many places, I can do a cleanup for this after #116712 fixed.

Comment (alculquicondor) [reply pkg/scheduler/eventhandlers.go:260]: you can wait... we are in code freeze anyways

Comment (AxeZhan) [reply pkg/scheduler/eventhandlers.go]: @alculquicondor @aojea I think `wait.PollUntilContextCancel` is the right function. PTAL
I can help replace the deprecated function `wait.Poll` in scheduler pkg in another pr, if it's ok.

File: pkg/scheduler/eventhandlers.go [hunk 20]
```diff
@@ -246,16 +249,39 @@ func responsibleForPod(pod *v1.Pod, profiles profile.Map) bool {
...
+	// syncedPollPeriod controls how often you look at the status of your sync funcs
+	syncedPollPeriod = 100 * time.Millisecond
+)
+
+// WaitForHandlersSync waits for EventHandlers to sync.
+// It returns true if it was successful, false if the controller should shut down
+func (sched *Scheduler) WaitForHandlersSync(ctx context.Context) error {
+	return wait.PollUntilContextCancel(ctx, syncedPollPeriod, true, func(ctx context.Context) (done bool, err error) {
+		for _, handler := range sched.registeredHandlers {
+			if !handler.HasSynced() {
```
Comment (alculquicondor) [inline pkg/scheduler/eventhandlers.go:262]: Can we have some v(3) that tell us which handlers are not synced? Ideally not for every iteration, but at least once per second.

Comment (AxeZhan) [reply pkg/scheduler/eventhandlers.go:262]: Current `registration` doesn't export a `Name` to us.
```
type ResourceEventHandlerRegistration interface {
	// HasSynced reports if both the parent has synced and all pre-sync
	// events have been delivered.
	HasSynced() bool
}

```
So in order to tell which handler is not synced, we need to create a new struct
```
type namedRegistration struct {
	cache.ResourceEventHandlerRegistration
	Name string
}
```
And create a name for every handler like:
```
handlers = append(handlers, namedRegistration{
	ResourceEventHandlerRegistration: handlerRegistration,
	Name:                             "dynInformer",
})
```
Is this way ok?

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve
/hold cancel

**Comment by AxeZhan** [comment]
kindly ping @aojea

**Comment by alculquicondor** [comment]
ping @aojea

File: cmd/kube-scheduler/app/server.go [hunk 21]
```diff
@@ -212,7 +212,11 @@ func Run(ctx context.Context, cc *schedulerserverconfig.CompletedConfig, sched *
 	if !cc.ComponentConfig.DelayCacheUntilActive || cc.LeaderElection == nil {
 		startInformersAndWaitForSync(ctx)
 	}
-
+	// Wait for all handlers to sync (all items in the initial list delivered) before scheduling.
+	if err := sched.WaitForHandlersSync(ctx); err != nil {
```
Comment (kerthcet) [inline cmd/kube-scheduler/app/server.go:216]: Shouldn't this be part of the `startInformersAndWaitForSync`? We'll also sync the cache in the leader elector mode.

**Comment by AxeZhan** [comment]
kindly ping @alculquicondor 
This pr is rebased due to some conflict with master, we need a "lgtm" again :)

File: pkg/scheduler/eventhandlers.go [hunk 22]
```diff
@@ -356,17 +392,26 @@ func addAllEventHandlers(
...
 			// Do nothing.
 		case framework.CSINode:
-			informerFactory.Storage().V1().CSINodes().Informer().AddEventHandler(
+			if handlerRegistration, err = informerFactory.Storage().V1().CSINodes().Informer().AddEventHandler(
 				buildEvtResHandler(at, framework.CSINode, "CSINode"),
-			)
+			); err != nil {
+				return err
+			}
+			handlers = append(handlers, handlerRegistration)
```
Comment (likakuli) [inline pkg/scheduler/eventhandlers.go:400]: There is no need to put it in the handlers, right?

**Pull request merged**